### PR TITLE
cross-container coordination demo

### PR DESCRIPTION
**Pull request by ransomw1c**
re #174 , this demo supplements the existing FUSE read-only demo by demonstrating cross-container coordination at the application level: conceptually, the system-level abstractions provided by kubernetes (volumes) are used as a transport layer to send events between applications in different containers within the same pod.

[here](https://github.com/kubernetes/kubernetes/issues/65502#issuecomment-415113553) is a comment from a kubernetes developer suggesting such bespoke intra-pod coordination of containers in regard to similar requirements on a different project.

### CODE REVIEW

**Comment by ransomw1c** [comment]
there's still a good bit of cleanup and commenting to be done, yet this demonstrates the basic idea:

`mock_application.sh` is a program that needs to access datamon, compute some results based on that data, and the upload it back to datamon.  both the datamon sidecar and the main container get their own respective `wrap_*.sh` scripts.  these wrappers send messages to one another by writing files to an `emptyDir` or other mechanisms provided by the system (using the `mount` command to determine whether datamon is ready for reading).  overall, it's a setup much like two communicating goroutines:  each wrapper is a sequential process that needs to send messages for the other process to receive before the other process unblocks.

**Comment by ransomw1c** [comment]
opening this for review..

.. the comments beginning with `##` in `wrap_datamon.sh` and `wrap_application.sh` describe the coordination flow in words, while the comments beginning with `#%` describe it using pseudo-go.

the question of how to parameterize at the kubernetes level in `example-coord.template.yaml` is still open.

**Comment by kerneltime** [comment]
The approach is good to start experimenting.

File: hack/k8s/example-coord.template.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        stdin: true
+        tty: true
+        volumeMounts:
+        - mountPath: /tmp/coord
+          name: container-coord
+      containers:
+      - name: demo-app
+        image: gcr.io/onec-co/datamon-fuse-demo-coord-app:latest
+        imagePullPolicy: "Always"
+        # todo: determine proper parameterization
```
Comment (ransomw1c) [inline hack/k8s/example-coord.template.yaml]: @galvare2 , these todos about the parameters to the `wrap_*.sh` a remaining uncertainty.

there's also the naming convention on the `volumeMounts`:  place them in `/tmp/`? `/`? `/datamon/`?  make the naming convention configurable via parameters to the script?

there are probably some other things to review here, yet despite the [wip] tag, this is ready for review when you have time.

File: hack/fuse-demo/wrap_datamon.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#! /bin/sh
+
+### datamon wrapper (demo)
+# half of the container coordination sketch, a script like this one
+# is meant to wrap datamon in the sidecar container of an Argo DAG node
+# and communciate with a script like wrap_application.sh.
+
+COORD_POINT=/tmp/coord
+UPLOAD_SOURCE=/tmp/upload
```
Comment (galvare2) [inline hack/fuse-demo/wrap_datamon.sh]: Will need a way to make this location configurable

**Comment by ransomw1c** [comment]
it turns out we need configurable arities and destinations on the mount (download) and upload paths.

thumbnail design sketchs:

##### `wrap_datamon.sh`

```
wrap_datamon.sh -u /x,/y -d bundlemount:repo=repo-a:label=label-a:mount=/a_mount:dest=/a_dest:stream,bundlemount:repo=repo-b:label=label-b:mount=/b_mount:dest=/b_dest:stream
```

or

```
wrap_datamon.sh -u /x -u /y -d bundlemount:repo=repo-a:label=label-a:mount=/a_mount:dest=/a_dest:stream,bundlemount:repo=repo-b:label=label-b:mount=/b_mount:dest=/b_dest:stream
```

(following the usual `foo:option1=value1:option2:option3=value3,bar` convention)

##### `wrap_application.sh`

```
wrap_application.sh  -u /x,/y -- <app>
```

or

```
wrap_application.sh  -u /x -u /y -- <app>
```

**Comment by ransomw1c** [comment]
as of, 35e2b6e , params are implemented as in

```
wrap_application.sh  -c /tmp/coord -- <app>
```

and

```
wrap_datamon.sh -c /tmp/coord -d 'bundle upload --path /tmp/coord --message "result of container coordination demo" --repo ransom-datamon-test-repo --label coordemo' -d 'bundle mount --repo ransom-datamon-test-repo --label testlabel --destination /tmp --mount /tmp/mount --stream'
```

in particular, there's not an additional parameter syntax to worry about:  just pass `bundle upload` and `bundle mount` parameter lists to the script as individual paramters, and it'll do the correct thing ... except if there are `:` chars in the parameter lists.

**Comment by ransomw1c** [comment]
in fact, there's an additional issue where parameters in datamon parameter lists (including the bundle upload message) can't have spaces in them... hopefully this will not be an issue with respect to getting some PoC usage started.

**Comment by ransomw1c** [comment]
hoping we can try this with some Argo workflows despite flaws mentioned in previous two comments..

..i'm converging on the idea of building much of the `wrap_datamon.sh` into the datamon binary itself, as `wrap_datamon.sh` is increasingly complex with the multi-arity requirement.

**Comment by ransomw1c** [comment]
to additionally consider:  allow a `config` command to init `~/.datamon/datamon.yaml` from `wrap_datamon.sh`.

**Comment by ransomw1c** [comment]
toggled this back to wip status in order to add documentation to the readme.

**Comment by ransomw1c** [comment]
> The application wrapper needs to pass through the exit code of the application it wraps.

**Comment by ransomw1c** [comment]
##### consolidated second revision changes
- [ ] allow a `config` command to init `~/.datamon/datamon.yaml` from `wrap_datamon.sh`.
- [ ] The application wrapper needs to pass through the exit code of the application it wraps.
- [ ] add documentation to the readme.

File: README.md [hunk 3]
```diff
@@ -111,6 +111,92 @@ with a label that already exists overwrites the commit hash previously associate
...
+where `<coordination_directory>` is an empty directory in a shared volume
+(an
+[`emptyDir`](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir)
+using memory-backed storage suffices).  In the case of Argo workflows in particular,
+the empty directory (and not necessarily the volume) ought to be specific to a
+particular DAG node (i.e. Kubernetes pod).  Each node uses a unique directory.
+And where `<application_command>` is the data-science executable as well as any
+parameters (such the paths to place data into or out of datamon, locations in volumes
+shared with the sidecar container) to that executable exactly as they would appear without
+the wrapper script.  That is, the wrapper script, relies the
```
Comment (galvare2) [inline README.md]: This sentence is worded confusingly

File: README.md [hunk 4]
```diff
@@ -111,6 +111,92 @@ with a label that already exists overwrites the commit hash previously associate
...
+[conventional UNIX syntax](http://zsh.sourceforge.net/Guide/zshguide02.html#l11)
+for stating that options to a command are done being declared.
+
+Meanwhile, `wrap_datamon.sh` similarly accepts a single `-c` option to specify the
+location of the coordination directory.  Reiterating: the coordination directory need
```
Comment (galvare2) [inline README.md]: This reiteration is probably not necessary

File: .circleci/config.yml [hunk 5]
```diff
@@ -42,6 +43,7 @@ jobs:
           name: Run golang linter
           command: |
             golangci-lint run --disable golint --build-tags fuse_cli --max-same-issues 0 --verbose
+            shellcheck hack/fuse-demo/wrap_*.sh
```
Comment (kerneltime) [inline .circleci/config.yml]: 👍

File: Makefile [hunk 6]
```diff
@@ -209,3 +177,62 @@ goimports:
...
+		--ssh default \
+		-f ./hack/fuse-demo/shell.Dockerfile \
+		.
+	docker push gcr.io/onec-co/datamon-fuse-demo-shell
+
+## demonstrate a fuse read-only filesystem
+fuse-demo-ro: fuse-demo-build-shell
+	@docker image push gcr.io/onec-co/datamon-fuse-demo-shell
+	@./hack/fuse-demo/create_ro_pod.sh
+	@sleep 8 # dumb timeout on container startup
```
Comment (kerneltime) [inline Makefile]: Not a blocker, you can take a look at [wait](https://hackernoon.com/kubectl-tip-of-the-day-wait-like-a-boss-40a818c423ac)

**Pull request merged**